units/boot-complete.target

```diff
@@ -0,0 +1,14 @@
+#  SPDX-License-Identifier: LGPL-2.1+
+#
+#  This file is part of systemd.
+#
+#  systemd is free software; you can redistribute it and/or modify it
+#  under the terms of the GNU Lesser General Public License as published by
+#  the Free Software Foundation; either version 2.1 of the License, or
+#  (at your option) any later version.
+
+[Unit]
+Description=Boot Completion Check
+Documentation=man:systemd.special(7)
+Requires=sysinit.target
+After=sysinit.target
```

Comment:
I think a very common use-case will be to unconditionnaly use boot-complete.target.

maybe add

[Install]
RequiredBy=default.target

so it is easy to do ? (of course, i'm not asking to enbale the target by default, just that the section be here for ease of use)

Reply:
What would the benefit of this be? If there's nothing ordered after the target that depends on it then all you get by this is some extra log output about a couple of failing services?

Reply:
maybe it's a missunderstanding on my side...

IIUC the way you see things, a user that wants to activate boot counting will enable systemd-boot-complete.service which will pull boot-complete.target right ?

It would make more sense to me to enable/disable boot-complete.target and have systemd-boot-complete.service have a WantedBy on it... This way the way to enable boot counting is bootloader-neutral.

Reply:
> IIUC the way you see things, a user that wants to activate boot counting will enable systemd-boot-complete.service which will pull boot-complete.target right ?

no. not quite. systemd-boot-complete.service is not supposed to be enabled. Instead, systemd-boot-complete-generator automatically pulls it into the initial transaction if it detects that boot counting is enabled.

Basically, to make use of all of this on your uefi system it is entirely sufficient to do `echo > /etc/kernel/tries`. The next kernel you install then will have boot counting enabled, and all will work fully automatically, systemd-boot will make use of it, and the generator ensures thes userspace side of this runs too.

> It would make more sense to me to enable/disable boot-complete.target and have systemd-boot-complete.service have a WantedBy on it... This way the way to enable boot counting is bootloader-neutral.

So my thinking was that there might be other reasons you'd want a boot complete check in place, not just boot counting stuff. For example, let's say you design a system where worker systems shall request a job to run from some central system if they booted up correctly, and only then. For usecases like that it might make sense to reuse boot-complete.target, even though this has very little to do with boot counting.

Reply:
well I don't use UEFI (well, I mean... I don't configure UEFI) so I don't really know the whole systemd-boot stuff. I don't really understand why you use a generator for boot counting instead of enabling boot counting statically or using a command-line parameter... Is it to have a way to enable it through UEFI variables ?

Anyway, if that's the idea I missed it in the doc. I have never seen a use-case where boot counting is dynamically disable. For all my use-cases it could be a compile option. (yes that's stupid for other reasons, but that would work for me)

Because I don't completely see what you are trying to do I probably miss the big picture, but I would like to avoid having to write a generator to replace yours for the embedded use-case, kowing it would just recreate a link every time that could be described with an [Install] section

As for enabling boot-complete.target vs enabling a systemd-boot specific service... my main grip is that's systemd-boot specific. so one more thing that i need to redo in a bootloader specific way...

Thing is... UEFI is an x86 only thing... (there was a move to implement it on ARM, but this seems to be dead by now, afaict) so we need to reimplement systemd-boot and its internal logic, which is not really documented. If there is a way to simplify that part, it's always better for us...

Reply:
A generator is preferable because it is "configurationless". Essentially, you decide boot-counting=yes/no before a kernel package is installed, and the rest happens automatically. We *could* ask the user to provide a proper switch somewhere at the kernel command-line, but setting the kernel command-line is fragile, and could get out of sync, etc. By auto-detecting if we need to do this, we make it very easy for the user, and also easier for the system installer. I don't think making a generator is so bad. Usually they are very short and simple, can be written in any language, and they are also easy to test.

Reply:
> well I don't use UEFI (well, I mean... I don't configure UEFI) so I don't really know the whole systemd-boot stuff. I don't really understand why you use a generator for boot counting instead of enabling boot counting statically or using a command-line parameter... Is it to have a way to enable it through UEFI variables ?

The only reason to have the generator is to make this less fragile. Instead of having to explicitly turn on boot counting in the boot loader and then also in systemd, it's entirely sufficient if you turn this on in the boot loader in a completely generic fashion, and systemd will pick it up.

My goal here was to make the coupling as lose but automatic as we can: everything gets detected automatically if the boot loader supports it, and all the files on disk actually remain compatible with boot loaders that do not support it.

Now, on UEFI we have the luxury of UEFI variables which can be checked very early on, before we mount any file systems. Very likely other boot systems do not have this luxury, and need to store their state exclusively in the file system. In that case a generator can't really work, and hence this PR cannot be used as 1:1 blueprint for that case. In that case it's probably advisible to instead make their systemd-boot-complete.service counterpart something you explicitly enable.

Reply:
well, it's more of the other way round on arm... there is usually a dedicated NAND sector for booloader variables, and we try to avoid relying on the filesystem at all... boot-counting is usually done in those variables rather than in a FAT partition which needs to be doubled for robustness purpose.

but I see where you are going. My use-case is usually simpler, since in the embedded world we don't care as much about automated deployment in the way datacenters do. For me kernel command line is good enough since we never deactivate boot counting.